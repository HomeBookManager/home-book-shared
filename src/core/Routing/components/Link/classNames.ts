export const className = 'Link';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      forcedHover: `${className}--forced-hover`,
      unerline: `${className}--underline`,
    },
  },
};
