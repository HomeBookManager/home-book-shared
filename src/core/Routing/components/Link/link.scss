@import '../../../../styles/colors/common.scss';
@import '../../../../styles/mixins/transitions/common.scss';

.Link {
  @include transition;
  color: $--hb-neutral-2;
  display: block;
  font-size: 14px;
  position: relative;
  text-decoration: unset;

  &:hover {
    color: $--hb-common-neutral-blue-1;
  }

  &--dark {
    color: $--hb-neutral-2--dark;

    &:hover {
      color: $--hb-common-neutral-blue-2;
    }
  }

  &--forced-hover {
    color: $--hb-common-neutral-blue-1;

    &--dark {
      color: $--hb-common-neutral-blue-2;
    }
  }

  &--underline {
    text-decoration: underline;
  }
}
