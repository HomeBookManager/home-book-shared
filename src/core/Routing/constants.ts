export const APPLICATION_NAME = 'HomeBook';

export const translationNameSpace = 'routing';
export const guardNameSpace = `${translationNameSpace}.guard`;
