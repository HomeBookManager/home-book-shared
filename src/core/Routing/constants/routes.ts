export enum RouteName {
  login = 'login',
  register = 'register',
}

export const ROUTES = {
  [RouteName.login]: '/login',
  [RouteName.register]: '/register',
};
