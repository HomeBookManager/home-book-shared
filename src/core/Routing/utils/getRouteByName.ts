export const getRouteByName = (
  name: string,
  routes: { [key: typeof name]: string },
): string => routes[name];
