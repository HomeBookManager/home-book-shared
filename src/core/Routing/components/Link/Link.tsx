import cx from 'classnames';
import { AnchorHTMLAttributes, FC, ReactNode } from 'react';
import { Link as RouterLink } from 'react-router-dom';

// components
import E2EDataAttribute from '../../../../components/E2EDataAttributes/E2EDataAttribute';

// hooks
import { useTheme } from '../../../../hooks/useTheme/useTheme';

// others
import { className as classNameLink, classNames } from './classNames';

// styles
import './link.scss';

// types
import { E2EAttribute } from '../../../../components/E2EDataAttributes/enums';

export type TProps = AnchorHTMLAttributes<HTMLAnchorElement> & {
  children?: ReactNode;
  e2eAttribute?: E2EAttribute;
  e2eValue?: number | string;
  forcedHover?: boolean;
  underline?: boolean;
};

const Link: FC<TProps> = ({
  children = null,
  className,
  e2eAttribute = E2EAttribute.link,
  e2eValue = '',
  forcedHover = false,
  href = '',
  target = '_self',
  underline = false,
  ...props
}) => {
  const opensInNewWindow = target === '_blank';
  const { classNamesWithTheme } = useTheme(classNames);

  const commonClassNames = cx(
    className,
    classNamesWithTheme[classNameLink].name,
    {
      [classNamesWithTheme[classNameLink].modificators.forcedHover]:
        forcedHover,
    },
    {
      [classNamesWithTheme[classNameLink].modificators.unerline]: underline,
    },
  );

  const isExternalLink = (href: string): boolean => {
    const externalLinkRegexp = /^https?:\/\//;
    return externalLinkRegexp.test(href);
  };

  if (isExternalLink(href) || opensInNewWindow) {
    return (
      <E2EDataAttribute type={e2eAttribute} value={e2eValue}>
        <a href={href} target={target} className={commonClassNames} {...props}>
          {children}
        </a>
      </E2EDataAttribute>
    );
  }

  return (
    <E2EDataAttribute type={e2eAttribute} value={e2eValue}>
      <RouterLink to={href} className={commonClassNames} {...props}>
        {children}
      </RouterLink>
    </E2EDataAttribute>
  );
};

export default Link;
