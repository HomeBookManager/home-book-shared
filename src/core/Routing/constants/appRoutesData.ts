// components
import LoginPage from '../../../sample/components/LoginPage/LoginPage';
import RegisterPage from '../../../sample/components/RegisterPage/RegisterPage';

// types
import { RouteName } from './routes';
import { TAppRoutesData } from '../types';

export const APP_ROUTES_DATA: TAppRoutesData = [
  {
    Component: LoginPage,
    name: RouteName.login,
  },
  {
    Component: RegisterPage,
    name: RouteName.register,
  },
];
