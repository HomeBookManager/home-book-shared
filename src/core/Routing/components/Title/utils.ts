// @home-book
import { TObject } from 'home-book-types';

// types
import { RouteName } from '../../constants/routes';

// utils
import { enumToArray } from '../../../../utils';
import { getRouteByName } from '../../utils/getRouteByName';

export const getTitlesTranslationKeys = (): TObject<string> =>
  enumToArray(RouteName).reduce(
    (obj, key) => ({
      ...obj,
      [getRouteByName(RouteName[key])]: `routing.title.${key}`,
    }),
    {},
  );
