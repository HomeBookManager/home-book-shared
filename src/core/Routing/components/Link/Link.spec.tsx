import { createMemoryHistory } from 'history';
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';

// components
import Link from './Link';

// types
import { RouteName } from '../../constants/routes';

// utils
import { getRouteByName } from '../../utils/getRouteByName';

const history = createMemoryHistory({
  initialEntries: [getRouteByName(RouteName.login)],
});

describe('Link snapshots', () => {
  it('should render as dynamic link', () => {
    // before
    const { asFragment } = render(
      <Router history={history}>
        <Link href="/">Link</Link>
      </Router>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render as static link', () => {
    // before
    const { asFragment } = render(
      <Router history={history}>
        <Link href="https://">Link</Link>
      </Router>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render without children', () => {
    // before
    const { asFragment } = render(
      <Router history={history}>
        <Link href="https://" />
      </Router>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should return with blank text', () => {
    // before
    const { asFragment } = render(
      <Router history={history}>
        <Link target="_blank">Link</Link>
      </Router>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
