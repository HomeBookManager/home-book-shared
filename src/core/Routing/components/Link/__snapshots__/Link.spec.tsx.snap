// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link snapshots should render as dynamic link 1`] = `
<DocumentFragment>
  <a
    class="Link"
    data-test-link=""
    href="/"
  >
    Link
  </a>
</DocumentFragment>
`;

exports[`Link snapshots should render as static link 1`] = `
<DocumentFragment>
  <a
    class="Link"
    data-test-link=""
    href="https://"
    target="_self"
  >
    Link
  </a>
</DocumentFragment>
`;

exports[`Link snapshots should render without children 1`] = `
<DocumentFragment>
  <a
    class="Link"
    data-test-link=""
    href="https://"
    target="_self"
  />
</DocumentFragment>
`;

exports[`Link snapshots should return with blank text 1`] = `
<DocumentFragment>
  <a
    class="Link"
    data-test-link=""
    href=""
    target="_blank"
  >
    Link
  </a>
</DocumentFragment>
`;
